src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseConcreteTypeAnalyzer.Collector.cs

```diff
@@ -0,0 +1,463 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the MIT license.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using Analyzer.Utilities.Extensions;
+using Analyzer.Utilities.PooledObjects;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.NetCore.Analyzers.Performance
+{
+    public partial class UseConcreteTypeAnalyzer
+    {
+        private sealed class Collector : IDisposable
+        {
+            private static readonly ObjectPool<Collector> _pool = new(() => new Collector());
+
+            public readonly ConcurrentDictionary<IFieldSymbol, bool> VirtualDispatchFields = new(SymbolEqualityComparer.Default);
+            public readonly ConcurrentDictionary<ILocalSymbol, bool> VirtualDispatchLocals = new(SymbolEqualityComparer.Default);
+            public readonly ConcurrentDictionary<IParameterSymbol, bool> VirtualDispatchParameters = new(SymbolEqualityComparer.Default);
+            public readonly ConcurrentDictionary<IMethodSymbol, bool> MethodsAssignedToDelegate = new(SymbolEqualityComparer.Default);
+
+            public readonly ConcurrentDictionary<IFieldSymbol, ConcurrentBag<ITypeSymbol>> FieldAssignments = new(SymbolEqualityComparer.Default);
+            public readonly ConcurrentDictionary<ILocalSymbol, ConcurrentBag<ITypeSymbol>> LocalAssignments = new(SymbolEqualityComparer.Default);
+            public readonly ConcurrentDictionary<IParameterSymbol, ConcurrentBag<ITypeSymbol>> ParameterAssignments = new(SymbolEqualityComparer.Default);
+            public readonly ConcurrentDictionary<IMethodSymbol, ConcurrentBag<ITypeSymbol>> MethodReturns = new(SymbolEqualityComparer.Default);
```

Comment:
IMO, we are still doing lots of allocations by allocating a `ConcurrentBag` for almost every member symbol in the compilation. Consider switching from `ConcurrentBag` to `PooledConcurrentSet` and allocating/freeing from the pool to reduce allocations.

Reply:
Done.